File: 114. Flatten Binary Tree to Linked List.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        
        helper(root);
        
    }
    
    private TreeNode helper(TreeNode root) {
        
        // base case
        if (root == null) {
            return null;
        }
        
        // take out the subtrees
        TreeNode left_subtree = root.left;
        TreeNode right_subtree = root.right;
        
        // delete the original left subtree
        root.left = null;
        
        // add the flattened left subtree to the right
        root.right = helper(left_subtree);
        
        // traverse to the end of our new tree
        TreeNode node = root;
        while (node.right != null) {
            node = node.right;
        }
        
        // add the flattened right subtree to the right
        node.right = helper(right_subtree);
        
        // return the root: tree is flattened
        return root;
        
    }
}

File: 114. Flatten Binary Tree to Linked List.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        def helper(root):
            # base case
            if not root:
                return None
            
            # take out subtrees
            left_subtree = root.left
            right_subtree = root.right
            
            root.left = None
            # add flattened left subtree to the right
            root.right = helper(left_subtree)
            # traverse to the end of new tree
            node = root
            while node.right:
                node = node.right
            # add flattened right subtree to the right
            node.right = helper(right_subtree)
            
            # flattened
            return root
        
        # flatten in-place
        helper(root)